### PR TITLE
[RFC] Support ARMv7 architectures

### PR DESCRIPTION
**Pull request by etienne-lms**
This P-R proposes to add support for ARMv7-A architectures in the ARM trusted firmware.

Obviously, only ARMv7 with the Trustzone extensions can be supported by ARM trusted firmware. The proposed changes aims at covering most such ARMv7 cores. However these changes where initially designed to cover a Cortex-A9 based system.

These changes are split in incrementing changes:
- a single change for basic ARMv7 support
- a proposal to support GICv1
- several changes for Cortex-A9 based systems (non LPAE, division, PL310)
- a late maybe a bit out of scope optimization regarding memory consumption of the xlat table on AArch32 only architectures)

A port for the Cortex-A9 based "b2260" board (96board form factor style) based on this RFC will be later proposed (available from branch armv7-b2260 from https://github.com/etienne-lms/arm-trusted-firmware)

### CODE REVIEW

File: docs/porting-guide.rst [hunk 1]
```diff
@@ -473,6 +473,36 @@ constants must also be defined:
    Defines the total size of the physical address space in bytes. For example,
    for a 32 bit physical address space, this value should be ``(1ull << 32)``.
```
Comment (danh-arm) [inline docs/porting-guide.rst]: This patch needs further discussion and should in any case be separate from the rest of this PR.

If memory footprint is the concern, then it should be possible for the platform to overlay BL32 RW data over BL2 so the MMU tables for those 2 images at least don't occupy in-use memory at the same time.

I can see your solution would share BL1 xlat table memory too but is there really much overall memory to be saved? The overall PLAT_BASE_XLAT_SIZE has to be large enough for the largest BL's needs. I expect this to be BL2, which is not resident at runtime. So the delta in the number of translation tables needed between BL2 and BL32 is wasted.

File: drivers/arm/pl310/pl310_asm.S [hunk 2]
```diff
@@ -0,0 +1,87 @@
+/*
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply drivers/arm/pl310/pl310_asm.S]: ok, understood.

File: include/lib/aarch32/arch.h [hunk 3]
```diff
@@ -8,6 +8,24 @@
 #define __ARCH_H__
 
 /*******************************************************************************
+ * Macros to identify ARMv7 extensions
+ ******************************************************************************/
+#define ARMV7_ARCH_MINOR_LPAE	(1 << 0)
+#define ARMV7_ARCH_MINOR_VE	(1 << 1)
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply include/lib/aarch32/arch.h]: Yes, I will propose something explicit, maybe from another place (maybe not arch.h).

File: include/lib/aarch32/arch.h [hunk 4]
```diff
@@ -102,15 +102,16 @@
 #define ID_PFR1_GIC_MASK	0xf
 
 /* SCTLR definitions */
-#define SCTLR_RES1	((1 << 23) | (1 << 22) | (1 << 11) | (1 << 4) | \
-			(1 << 3))
+#define SCTLR_RES1	((1 << 23) | (1 << 22) | (1 << 4) | (1 << 3))
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply include/lib/aarch32/arch.h]: Ok. Maybe also for the armv7s where branch prediction enable is controlled by mmu state?

File: include/lib/aarch32/smcc_macros.S [hunk 5]
```diff
@@ -20,6 +20,44 @@
 	mov	r0, sp
 	add	r0, r0, #SMC_CTX_SP_USR
 
+#if ARCH_IS_ARMV7_WITHOUT_VE
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply include/lib/aarch32/smcc_macros.S]: ok, understood.

File: include/lib/xlat_tables/xlat_tables_defs.h [hunk 6]
```diff
@@ -7,6 +7,7 @@
 #ifndef __XLAT_TABLES_DEFS_H__
 #define __XLAT_TABLES_DEFS_H__
 
+#include <arch.h>
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply include/lib/xlat_tables/xlat_tables_defs.h]: No good reason but time...

File: lib/psci/psci_setup.c [hunk 7]
```diff
@@ -259,8 +259,10 @@ int psci_setup(const psci_lib_args_t *lib_args)
  ******************************************************************************/
 void psci_arch_setup(void)
 {
+#if !ARCH_IS_ARMV7_WITHOUT_VE
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply lib/psci/psci_setup.c]: +1

File: lib/xlat_tables/aarch32/nonlpae_tables.c [hunk 8]
```diff
@@ -0,0 +1,531 @@
+/*
+ * Copyright (c) 2016-2017, Linaro Limited. All rights reserved.
+ * Copyright (c) 2014-2017, ARM Limited and Contributors. All rights reserved.
+ * Copyright (c) 2014, STMicroelectronics International N.V.
+ * All rights reserved.
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply lib/xlat_tables/aarch32/nonlpae_tables.c]: ok. I'll go for option 2 when possible. thanks.

**Comment by etienne-lms** [comment]
Closing P-R. Replaced with https://github.com/ARM-software/arm-trusted-firmware/pull/1145.

**Pull request closed without merging**